### DISCUSSION TITLE
Project timeline (Mermaid Gantt diagram) generator

### DISCUSSION THREAD

**Discussion by Jajaho** [hugomods/.github#11] [Ideas]
Hello gals & guys!

I'm new to the party and just tried out my first hugoblox theme. It's very impressive to me, and I want to thank all the contributors for their hard work and dedication. 

## The Idea

Grab the start and end date from the front matter of project posts to generate a [Mermaid Gantt diagram](https://mermaid.js.org/syntax/gantt.html) when deploying. 
Sections of the diagram could be project categories (e.g. software, woodworking, ...) and projects with only a start date could be drawn with a fixed size. Project without a date omitted. 
The digram could then be put on the landing page of the project menu within a block.

## Implementation

I have some C, Rust and java experience but never written anything in Go, so I would be gracious for some pointers on how to best implement it in the language and especially how to make it fit in with hugo and the hugoblox infrastructure. 
I would guess that it is probably a good idea to make it a hugo module, which are based on Go modules.

## Questions

1. Where is the place to trigger the generation of the diagram, maybe use hooks?
2. What would be the suitable place/folder to output the generated diagram `public` or better `resources\_gen\`?
3. Do you have any guides or examples you could point me to?

Thank you very much for your time and help.

Best,
Jakob

**Comment by razonyang** [comment]
> Where is the place to trigger the generation of the diagram, maybe use hooks?

Hi, I have no experience with hugoblox. You may need to refer to their documentation for inserting code into pages.

> What would be the suitable place/folder to output the generated diagram public or better resources\_gen\?

I don't think there is such option in Hugo itself, generated files should be in `public` (by default) folder, the `resources` folder is more likely used by Hugo internally (caches).

> Do you have any guides or examples you could point me to?

It seems that the hugoblox has implemented Mermaid, see https://docs.hugoblox.com/reference/markdown/#diagrams.

This idea is likely limited to that theme. I suggest asking this on [their discussions](https://github.com/HugoBlox/hugo-blox-builder/discussions).

## Some hints

Unlike the backend language, the static site/frontend probably uses JavaScript to render diagrams on the browser side, so the key points are as follows.

1. Tell theme to load the Mermaid script.
2. Use hook (or sth similar) to inject the raw diagram definition into pages.

Since hugoblox has its implementation, you'll need to ask about it or figure it out through their documentation and codes.
